src/hotspot/share/gc/g1/g1VMOperations.cpp

```diff
@@ -125,33 +127,40 @@ VM_G1CollectForAllocation::VM_G1CollectForAllocation(size_t         word_size,
 void VM_G1CollectForAllocation::doit() {
   G1CollectedHeap* g1h = G1CollectedHeap::heap();
 
-  if (_word_size > 0) {
-    // An allocation has been requested. So, try to do that first.
-    _result = g1h->attempt_allocation_at_safepoint(_word_size,
-                                                   false /* expect_null_cur_alloc_region */);
-    if (_result != nullptr) {
-      // If we can successfully allocate before we actually do the
-      // pause then we will consider this pause successful.
-      _gc_succeeded = true;
-      return;
+  // Any allocation requests that were handled during a previous GC safepoint but have not been observed
+  // by the requesting mutator thread should be reset to pending. This makes it easier for the current GC to
+  // treat the unclaimed memory as garbage.
```

Comment:
This would create a dependence between handle_allocation_requests, and any collections that happen before. So we would have to deal with how these `fillerObjects` are treated by the collections, some could be invalidated.

Reply:
In the general case this is true, but here, before GC occurred, the existing allocations can still be considered valid until we know that we need a gc.
Doing so potentially avoids garbage collections because then we are not going to throw away all of them and try to reallocate.

I.e. the suggestion is like:
```
// keep existing allocations, they are still valid here.

if (has_allocations) {
  if (try_satisfy_allocations()) {  // satisfy non-satisfied allocations.
    return; // no gc occurred, both the previously satisfied allocations and the new allocations are valid.
  }
}
// We are about to do a GC. Reset all allocation requests since we are likely going to free regions containing them.
reset_allocation_requests();

// do gc
```
instead of the current code
```
reset_allocation_requests(); // throw away all requests, even already satisfied ones

if (has_allocations) {
  if (try_satisfy_allocations()) { // satisfy all allocations again
    return; // done, all satisfied (no gc occurred)
  }
  reset_allocation_requests(); // throw away all requests, prepare for GC
}
```

Reply:
Agreed!